core/context.go

```diff
@@ -0,0 +1,222 @@
+type Context struct {
+	users  []User
+	groups []gouser.Group
+	shells []string
+}
+
+// User is a convenience struct with information on a user entry in /etc/passwd
+type User struct {
+	Name     string
+	Groups   []string
+	IsActive bool
+	IsRoot   bool
+	IsAdmin  bool
+}
+
+// NewContext creates an initialized instance of a Context object
+func NewContext() (ctx *Context, err error) {
+	ctx = &Context{}
+	log.Debugln("Gathering system info...")
+
+	ctx.shells = activeShells()
+	log.Debugln("\tGathered active shells from /etc/shells")
+
+	if err = ctx.populateGroups(); err != nil {
+		return ctx, fmt.Errorf("failed to obtain groups from /etc/groups: %s", err)
+	}
+	log.Debugln("\tGathered groups from /etc/groups")
+
+	if err = ctx.populateUsers(); err != nil {
+		return ctx, fmt.Errorf("failed to obtain users from /etc/passwd: %s", err)
+	}
+	log.Debugln("\tGathered users from /etc/passwd")
+
+	return ctx, err
+}
+
+// FilterUsers returns users that fit into one or more of the given filters
+func (c *Context) FilterUsers(filters ...string) (filtered []User) {
+	for _, it := range c.users {
+		switch {
+		case contains(filters, "all"):
+			fallthrough
+		case it.IsActive && contains(filters, "active"):
+			fallthrough
+		case !it.IsActive && contains(filters, "system"):
+			fallthrough
+		case (it.IsRoot || it.IsAdmin) && contains(filters, "admin"):
+			filtered = append(filtered, it)
+		}
+	}
+
+	return filtered
+}
+
+// AddToGroup adds a User to a preexisting group. Returns whether or not the modification ran, along with an error in
+// case something went wrong
+func (c *Context) AddToGroup(user *User, group string) (ran bool, err error) {
+	if !contains(user.Groups, group) {
+		cmd := exec.Command("usermod", "-aG", group, user.Name)
+		if err = cmd.Run(); err != nil {
+			return false, err
+		}
+
+		user.Groups = append(user.Groups, group)
+		ran = true
+	}
+	return ran, err
+}
+
+// CreateGroup creates a new group with the given name and ID
+func (c *Context) CreateGroup(name string, id string) error {
+	cmd := exec.Command("groupadd", "-g", id, name)
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	c.groups = append(c.groups, gouser.Group{
+		Name: name,
+		Gid:  id,
+	})
+
+	return nil
+}
+
+// UpdateGroupID finds a group with the given name and changes its ID to the given ID
+func (c *Context) UpdateGroupID(name string, id string) error {
+	cmd := exec.Command("groupmod", "-g", id, name)
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	for _, it := range c.groups {
+		if it.Name == name {
+			it.Gid = id
+			break
+		}
+	}
+
+	return nil
+}
+
+func (c *Context) populateGroups() (err error) {
+	c.groups = make([]gouser.Group, 0)
+
+	C.setgrent()
+	for {
+		gr := C.getgrent()
+		if gr == nil {
+			break
+		}
+
+		group, err := gouser.LookupGroup(C.GoString(gr.gr_name))
+		if err != nil {
+			return err
+		}
+
+		c.groups = append(c.groups, *group)
+	}
+	C.endgrent()
+
+	return err
```

Comment:
Can just be `return` since err is a named variable being returned in `(err error)`